File: examples/rust/cond-write-multi-field/CondWriteMultiField.runar.rs
use runar::prelude::*;

/// CondWriteMultiField -- regression fixture for GitHub issue #99: a
/// conditional write of two mutable state fields in an `if` without an `else`.
#[runar::contract]
pub struct CondWriteMultiField {
    pub a: Bigint,
    pub b: Bigint,
}

impl CondWriteMultiField {
    pub fn bump(&mut self, flag: Bigint) {
        if flag > 0 {
            self.a = self.a + 1;
            self.b = self.b + 2;
        }
        self.add_output(1000, self.a, self.b);
    }
}

File: examples/zig/cond-write-multi-field/CondWriteMultiField.runar.zig
const runar = @import("runar");

pub const CondWriteMultiField = struct {
    pub const Contract = runar.StatefulSmartContract;

    a: i64 = 0,
    b: i64 = 0,

    pub fn init(a: i64, b: i64) CondWriteMultiField {
        return .{ .a = a, .b = b };
    }

    pub fn bump(self: *CondWriteMultiField, flag: i64) void {
        if (flag > 0) {
            self.a = self.a + 1;
            self.b = self.b + 2;
        }
        self.addOutput(1000, self.a, self.b);
    }
};
